src/btree/bt_cursor.c

```diff
@@ -2296,14 +2299,56 @@ __wt_btcur_bounds_position(
 
     if (CUR2BT(cursor)->type == BTREE_ROW) {
         WT_RET(__cursor_row_search(cbt, false, NULL, NULL));
-        /*
-         * If there's an exact match, the row-store search function built the key in the cursor's
-         * temporary buffer.
-         */
-        WT_RET(__wt_cursor_valid(cbt, (cbt->compare == 0 ? cbt->tmp : NULL), WT_RECNO_OOB, &valid));
+        if (cbt->compare == 0) {
+            /*
+             * If our compare value is 0 we found the exact key we were looking for, if the bound we
+             * are searching for is inclusive we can return it without walking.
+             */
+            if (F_ISSET(
+                  cursor, next ? WT_CURSTD_BOUND_LOWER_INCLUSIVE : WT_CURSTD_BOUND_UPPER_INCLUSIVE))
+                WT_RET(__wt_cursor_valid(cbt, cbt->tmp, WT_RECNO_OOB, &valid, false));
+            else
+                *need_walkp = true;
+        } else if (cbt->compare > 0) {
+            /* We landed to the right of search key. */
+            if (next)
+                /*
+                 * We need to check if the upper bound is being honored. We could optimize the call
+                 * to cursor valid and have it only check a single bound, but we would need to
+                 * indicate the bound and direction.
+                 */
+                WT_RET(__wt_cursor_valid(cbt, NULL, WT_RECNO_OOB, &valid, true));
+            else
+                /*
+                 * We know that this isn't a valid as we're doing a bound position call on the upper
```

Comment:
"this isn't a valid" should be "this isn't valid"

Reply:
[6096b21](https://github.com/wiredtiger/wiredtiger/pull/8309/commits/6096b21b1f569d91c83d5af8d2fafce7effe577b)